[CS] Don't apply compatibility logic to collection literals

hamishknight opened the pull request:
We currently permit (but warn on) coercions of collections with unrelated element types in certain cases. This is done in an effort to preserve compatibility with pre-5.3 compilers that may have allowed such code to compile due to dropping constraints while solving (#30886).

This is limited to collection coercions as we emit somewhat reasonable code for them that performs a force cast of the elements at runtime. However, it turns out that in the case of collection literals, we peephole the element conversions in CSApply, leading to codegen crashes. As such, narrow down the condition of this compatibility logic to not apply to collection literals, as this never would have compiled properly.

Additionally, disable the compatability logic altogether in Swift 6 mode.

rdar://88334481

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

hamishknight commented:
@swift-ci please test

hamishknight commented:
@swift-ci please test source compatibility

xedin commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -9555,16 +9555,34 @@ ConstraintSystem::simplifyBridgingConstraint(Type type1,
+    if (!rawType1->lookThroughAllOptionalTypes()->isTypeVariableOrMember())
+      return false;
+
+    auto *loc = getConstraintLocator(locator);
```
Nit: It's unnecessary to materialize locator of the checks your are doing here. Use `getLocatorParts` instead.

hamishknight committed:
[CS] Simplify visitCoerceExpr slightly

Remove the unused `choice` parameter.

hamishknight committed:
[CS] Don't apply compatibility logic to collection literals

We currently permit (but warn on) coercions of
collections with unrelated element types in
certain cases. This is done in an effort to preserve
compatibility with pre-5.3 compilers that may have
allowed such code to compile due to dropping
constraints while solving.

This is limited to collection coercions as we
emit somewhat reasonable code for them that
performs a force cast of the elements at runtime.
However, it turns out that in the case of collection
literals, we peephole the element conversions in
CSApply, leading to codegen crashes. As such,
narrow down the condition of this compatibility
logic to not apply to collection literals, as this
never would have compiled properly.

rdar://88334481

hamishknight committed:
[CS] Don't apply compatibility logic in Swift 6 mode

This was only intended to preserve compatibility
for cases where the constraint system may have
previously accepted invalid code. Drop it in Swift 6
mode such that invalid collection coercions become
errors.

slavapestov commented:
Can you disable the compatibility hack altogether in -swift-version 6 mode?

hamishknight commented:
@slavapestov Yup, 04b37ae1d0d87540d7254e0cf1c5e23988a207bd does that. Let me add it to the PR description
